### PR TITLE
add solutions to assignment 4

### PR DESCRIPTION
**Pull request by katermar**

### CODE REVIEW

File: assignment4/IslandCounter.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        for (int i = 0; i < row; i++) {
+            for (int j = 0; j < column; j++) {
+                // If cell water we don't need to union it with something
+                if (a[i][j] == 'F') {
+                    continue;
+                }
+                int current = i * column + j;
+                //check 4 neighbors and union, if it is land
+                if (i + 1 < row && a[i + 1][j] == 'T') {
+                    disjointSet.union((i + 1) * column + j, current);
```
Comment (brzozan) [inline assignment4/IslandCounter.java]: The calculation of they key to be used in disjoint set is repeated multiple times. It would be good to extract it to a separate method.

File: assignment4/IslandCounter.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
+package com.katermar.assignment4;
+
+/**
+ * Created by katermar on 6/21/2017.
+ */
+public class IslandCounter {
+    private char a[][];
```
Comment (brzozan) [inline assignment4/IslandCounter.java]: A more descriptive name would be nice... :)

Also field can be final.

File: assignment4/IslandCounter.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
+package com.katermar.assignment4;
+
+/**
+ * Created by katermar on 6/21/2017.
+ */
+public class IslandCounter {
+    private char a[][];
+
+    public IslandCounter(char[][] a) {
```
Comment (brzozan) [inline assignment4/IslandCounter.java]: As per assignment input is a 2d array of booleans rather than chars. If you accept chars, you have to worry about someone putting 'x' or any other character there.

File: assignment4/IslandCounter.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        this.a = a;
+    }
+
+    /*counts number of islands in an array
+    @return number of islands
+    */
+    public int countIslands() {
+        int row = a.length;
+        int column = a[0].length;
```
Comment (brzozan) [inline assignment4/IslandCounter.java]: What if someone passes `{}` as map?

File: assignment4/IslandCounter.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+                }
+                if (j + 1 < column && a[i][j + 1] == 'T') {
+                    disjointSet.union(i * column + (j + 1), current);
+                }
+                if (j - 1 >= 0 && a[i][j - 1] == 'T') {
+                    disjointSet.union(j * column + (j - 1), current);
+                }
+            }
+        }
+        int[] frequencyOfSets = new int[column * row];
```
Comment (brzozan) [inline assignment4/IslandCounter.java]: Any other data structure could be used to detect number of unique islands?

File: assignment4/IslandCounterTest.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        char emptyArray[][] = {{}, {}};
+        testIslandCounter(3, map);
+        testIslandCounter(2, map);
+        testIslandCounter(0, emptyArray);
+    }
+
+    public static void testIslandCounter(int expectedResult, char map[][]) {
+        char checkEmpty[][] = new char[map.length][map[0].length];
+        if (Arrays.deepEquals(map, checkEmpty)) {
+            System.out.println("ERROR! Array is EMPTY! Test FAILED!");
```
Comment (brzozan) [inline assignment4/IslandCounterTest.java]: Why? Assignment does not forbid empty array. Also the array you use as empty is not really empty. It is one of edge conditions, but there is another one: `{}`.

Regarding checking if array is empty, why not just check lengths rather than comparing with an object?

File: assignment4/IslandCounterTest.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+public class IslandCounterTest {
+
+    public static void main(String args[]) {
+        char map[][] = {{'F', 'T', 'F', 'T'},
+                        {'T', 'T', 'F', 'F'},
+                        {'F', 'F', 'T', 'F'},
+                        {'F', 'F', 'T', 'F'}};
+        char emptyArray[][] = {{}, {}};
+        testIslandCounter(3, map);
+        testIslandCounter(2, map);
```
Comment (brzozan) [inline assignment4/IslandCounterTest.java]: The second test always fails. The PR should contain only passing tests, so that they can be used to easily check if code works.
